lib/cli/kit/system.rb

```diff
@@ -191,9 +191,15 @@ def resolve_path(a, env)
           # If only one argument was provided, make sure it's interpreted by a shell.
           return ["true ; " + a[0]] if a.size == 1
           return a if a.first.include?('/')
-          item = env.fetch('PATH', '').split(':').detect do |f|
-            File.exist?("#{f}/#{a.first}")
+
+          paths = env.fetch('PATH', '').split(':')
+          item = paths.detect do |f|
+            command_path = "#{f}/#{a.first}"
+            next false unless File.exist?(command_path)
+            stat = File.stat(command_path)
```

Comment:
stat is super slow, maybe too slow for this frequent loop.

```rb
require 'benchmark'

n=50000

Benchmark.bmbm(15) do |x|
  %w(/ null).each do |f|
    x.report("file: #{f}") do
      n.times do
        next unless File.exist?(f)
        File.executable?(f)
        File.file?(f)
      end
    end

    x.report("file+stat: #{f}") do
      n.times do
        next unless File.exist?(f)
        stat = File.stat(f)
        stat.executable?
        stat.file?
      end
    end

    x.report("stat+rescue: #{f}") do
      n.times do
        begin
          stat = File.stat(f)
          stat.executable?
          stat.file?
        rescue => e
        end
      end
    end
  end
end
```

```
Rehearsal -----------------------------------------------------
file: /             0.250000   0.250000   0.500000 (  0.505493)
file+stat: /        1.300000   0.990000   2.290000 (  7.428403)
stat+rescue: /      1.310000   0.940000   2.250000 (  7.641827)
file: null          0.060000   0.040000   0.100000 (  0.106408)
file+stat: null     0.070000   0.030000   0.100000 (  0.101721)
stat+rescue: null   0.250000   0.060000   0.310000 (  0.367537)
-------------------------------------------- total: 5.550000sec

                        user     system      total        real
file: /             0.250000   0.250000   0.500000 (  0.500268)
file+stat: /        1.280000   0.970000   2.250000 (  7.253366)
stat+rescue: /      1.210000   0.870000   2.080000 (  7.035531)
file: null          0.060000   0.030000   0.090000 (  0.093907)
file+stat: null     0.060000   0.040000   0.100000 (  0.095574)
stat+rescue: null   0.210000   0.040000   0.250000 (  0.260318)
```

More than twice as fast to use `File.*` 🤷‍♂️ 
My theory is because stat gathers much more info.

Reply:
Interesting! Thanks for benchmarking, will definitely make these changes 💪

Reply:
Dug a tiny bit. They both call `rb_stat`, so it's likely due to the creation/allocation of a `File::Stat` object.